### PR TITLE
Command dedup: add columns to completions in append-only schemas [KVL-1057]

### PR DESCRIPTION
**Pull request by fabiotudone-da**
## Description

This is a further change in a series (after #10619) to let the Ledger API provide more command de-duplication information in completions as well (only the submission rank is deferred to later stages), so that the application can use it in correlation with the one provided at command submission time.

It updates the index DB append-only schema to allow storing the additional command de-duplication information and provides the read/write logic for it.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by fabiotudone-da** [comment]
@digital-asset/kv-participant Could you please have a first stab at the schema changes?

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql:10]: I guess these two will need to be nullable to accomodate historical data.

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql:10]: Yes, absolutely. And I recommend making them nullable on all databases so they're not out of sync.

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:126]: There is already an `application_id` above.

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql:8]: The H2 and Oracle append-only index databases already have an `application_id`. Are you sure that the PostgreSQL index DB doesn't have this column already?

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql:10]: Will this work? How do we populate for old data?

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql]: 8079673

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql]: 8079673

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql]: Made both nullable in all the schemas: 3e5fa10

**Review by fabiotudone-da — COMMENTED** [review]
A couple of additional things I am not sure about:

- Do we have a policy for the position of newly added columns?
- I couldn't find another instance of duration, is the double i64+i32 column (as in protobuf) OK?

**Comment by nmarton-da** [comment]
> A couple of additional things I am not sure about:
> 
>     * Do we have a policy for the position of newly added columns?
AFAIK no policy regarding position ATM
> 
>     * I couldn't find another instance of the duration, is the double i64+i32 column (as in protobuf) OK?
Do you need such precision? Wouldn't be seconds or millis enough (resulting in one bigint column)?
Besides no idea what datatype best to store duration, I would go with millis

**Comment by nmarton-da** [comment]
General comment to this PR
Not sure about you intentions, but it would be great (most probably necessary) to make the DB changes together with the DB facing code changes:
* Reviewers would gain more clarity where the schema is heading semantically
* It would be clear how backwards compatibility is ensured
* It would probably also include db facing tests, which would ensure smooth operation between schema and code
* It would successfully prevent possible multiple iterations on the schema due to integration with code (indexes, data-type changes, constraint changes, migration changes etc)

**Comment by fabiotudone-da** [comment]
> General comment to this PR

@nmarton-da Thanks for the feedback; the intention was to break down the work in small, separate reviews as much as possible, starting with the ones requiring the most feedback (like the ledger API proto and the DB schema, the latter purely on a technical level initially).
I think creating additional PRs on top of this one with the code changes (and waiting for a review on them too before merging anything) could have the advantages of both approaches. WDYT?

**Comment by fabiotudone-da** [comment]
> Do you need such precision? Wouldn't be seconds or millis enough (resulting in one bigint column)?
Besides no idea what datatype best to store duration, I would go with millis/bigint

**Comment by nmarton-da** [comment]
> > General comment to this PR
> 
> @nmarton-da Thanks for the feedback; the intention was to break down the work in small, separate reviews as much as possible, starting with the ones requiring the most feedback (like the ledger API proto and the DB schema, the latter purely on a technical level initially).
> I think creating additional PRs on top of this one with the code changes (and waiting for a review on them too before merging anything) could have the advantages of both approaches. WDYT?

I think the schema change and the code which works with that should be same unit in terms of work / mergable / PR.
How about keeping this a draft, after agreement reached, closing it, and then open another one which is about the change of the persistence with this content, and with the code changes?

Don't get me wrong please, small PRs is a very good thing, but there must be a PR which is too small :) I claim making db changes without the supporting code falls into the too small category. Also don't get me wrong please on the other end neither: I am not saying that because that should belong together with something else, then everything needs to belong together: feel free to maintain granular changes later as you see fit!

TLDR:
I would applaud if we could have a selfcontained PR:
* having db schema changes
* having StorageBackend code changes
* having JdbcLedgerDao code changes if needed
* and db facing tests proving successful SQL-scala integration
...together in one PR

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala:280]: I feel like I am hijacking a couple of test cases here but I need to check all the current dedup cases on a rejection and a successful completion (I don't really care which transaction). Let me know if you can think of a better way.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala:252]: Could extract these together and bind-deconstruct them earlier, maybe slightly better.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CompletionStorageBackendTemplate.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CompletionStorageBackendTemplate.scala:88]: These are the newly added (+ pre-existing application ID) command dedup columns, they will be used in the next PR.
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CompletionStorageBackendTemplate.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CompletionStorageBackendTemplate.scala:106]: These are the newly added (+ pre-existing application ID) command dedup columns, they will be used in the next PR.

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:138]: This way of specifying the deduplication period is currently unsupported but it belongs here for imminent use, unsure if removing it or leaving it for posterity.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala:241]: You could maybe do something like:

```scala
      deduplication_offset = deduplication.leftOr(None)(Some),
      deduplication_time_seconds = deduplication.middleOr(None)(Some),
      deduplication_time_nanos = deduplication.rightOr(None)(Some),
```

That said, I think I'd prefer a custom set of classes, with appropriate getters.

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:126]: I agree with leaving it and keeping the migration count down.

**Review by SamirTalwar-DA — COMMENTED** [review]
Looks pretty good to me, but I'll leave approval to the index maintainers.

**Comment by SamirTalwar-DA** [comment]
> Good point, I couldn't find any mention of resolution restrictions w.r.t. the API type (i.e., protobuf's). @SamirTalwar-DA Do you know anything more here?

I highly suggest not throwing away information at this point. If you want to store the value in one column, with a different precision, I'd recommend doing that before it makes it to the ledger.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala:241]: It would become even more cryptic:

```
      deduplication_offset = deduplication.flatMap(_.leftOr(None[_])(Some(_))),
      deduplication_time_seconds = deduplication.flatMap(_.middleOr(None[_])(Some(_))).map(_._1),
      deduplication_time_nanos = deduplication.flatMap(_.middleOr(None[_])(Some(_))).map(_._2),
```

I think I'll go with an ADT.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala:208]:
```suggestion
  sealed trait Deduplication extends Product {
```

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:109]: discussions on the precision made me think more about this:
* Foreword: sorry to raise this so late, but this is exactly what is my main concern regarding too granular PRs: not knowing the whole picture makes it quite hard to reason about this
* It seems to me that all the old fields are pass-through info for the indexer (except from the offset). How about the new ones? Does the indexer need to be aware of all this info? Can't we just squeeze all the non-relevant fields into a proto which could be possibly passed through the system? (this pattern can be captured in all the contract related serialized data and also in here since @SamirTalwar-DA 's commit, the rejection_status_details)

**Review by nmarton-da — COMMENTED** [review]
* The test for the low level database layer is located in StorageBackendSuite
* If you would want to test with that test suite, you would require to change the CompletionStreamResponse, as I suggested, otherwise you won't be able to observe outcome. (whether CompletionStreamResponse is both a persistent and API data-type is good or not, it debatable. BTW this is also a sign which made me think about making all this info a pass-through for the indexer/participant-api)
* So if you would have wanted to organize you're changes in PRs, and keep the persistence code tested the same time you would need to follow a different grouping: first doing db-schema changes, Dto, Schema, CompletionStreamResponse, and test then can be added to SotrageBackendSuite, and only then doing the UpdateToDbDto changes, with unit tests (which actually would unlock JdbcLedgerDao related testing, and integration testing as well)

File: ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:109]: My general idea is that everything that needs to be serialized (e.g. b/c it's driver-dependent) should be, but nothing more because that makes using the info more complicated.

In this case, it's the new command dedup model and it's API behavior, so I think we shouldn't make it opaque at the index level and it's supposed not to change in the (near) future; for this reason, maintenance should also be minimal after this initial work.

Being core info, it's also possible that the indexer may benefit from being aware of these new fields in the future (although the design trend for command deduplication implementations is currently driver-based). In order to make it easy to use (and to migrate) them in future designs/architectures, I think it's better if they are known to the DB.
Another use case is debugging/troubleshooting; it is true that there are other means (e.g. logging) but being able to look at the actual data in the DB at a given point in time can be pretty useful in some circumstances.

I am aware that even just adding pass-through fields is a lot of work and code but maybe that's the actual issue and we should rather deal with that.

Does it make sense?

Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql:109]: After discussion with @meiersi-da and @rautenrieth-da the concolusion is that it is okay in this case to keep this pass-thru data flattened :ok_hand:

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala]: > wouldn't it will be simpler just to extend dtoCompletion with the default null-s, and override the needed ones in the test

Yes and necessary to test for unexpected data. Done so.

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql]: Actually, application ID was not nullable in the PSQL schema, I was thrown off by a comment there. Reverted to non-nullable in all schemas.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala:44]: Can you make this a `Ref.SubmissionId`? All other `some*` values are properly typed as well.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala:266]: Maybe you could have left `"handle TransactionAccepted (single create node)"` alone and added a new test `"handle TransactionAccepted (all deduplication data)"`. This would make it more clear what the purpose of each test is, at the cost of duplicated code.

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala:126]:
```suggestion
    // The only invariant that should hold, considering legacy data, is that either
```

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestValues.scala]: 87da1dd

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala]: f32346e

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDbDto.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDbDto.scala]: 482891e and  804c99e

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V108_add_deduplication_info_to_completions.sql]: (More correctly, it is still nullable because it also contained checkpoints but they were migrated out at some point)

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala:266]: Yes, I think that test purpose clarity is more important so I will do as you suggest before merging.

Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala]: bde3d1e, there's a comprehensive code deduplication work that can be done but I'd rather do it in a separate future PR.

**Review by nmarton-da — APPROVED** [review]
Looks solid.
Thanks for your patience regarding the back and forth-s

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDbDto.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDbDto.scala]: 4a94562

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala]: It's mostly because of scalapb plus proto's "non-optional optionals"; I simplified it further in 2749c75 and I think a unit test at this level doesn't add much value because it's very thin logic and the test would basically duplicate it in order to check it. I think testing the `CompletionFromTransaction` unit at its interface (i.e., `acceptedCompletion` and `rejectedCompletion`) is in a similar situation as it just delegates to the private `toXXX` utilities.

Also, this logic should  already have full coverage by the DB-level integration test cases in `StorageBackendTestsCompletions`. WDYT?

Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala]: I'll merge this as soon as the build passes and we'll deal with further improvements in a separate PR, if need be.

**Pull request merged**
